src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,181 @@
+                size++;
+                return;
+            }
+            Node<T> currentNode = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexIsValid(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexIsValid(index);
+        T temp = getNodeByIndex(index).item;
+        getNodeByIndex(index).item = value;
+        return temp;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexIsValid(index);
+        /* if we need remove single Node */
+        if (size == 1 && index == 0) {
+            size--;
+            Node<T> removedNode = tail;
+            tail = head = null;
+            return removedNode.item;
+        }
+        /* if we need remove Node from the end */
+        if (index == size - 1) {
+            size--;
+            Node<T> removedNode = tail;
+            tail = tail.prev;
+            tail.next.prev = tail.next = null;
+            return removedNode.item;
+        }
+        /* if we need remove Node from the beginning*/
+        if (index == 0) {
+            size--;
+            Node<T> removedNode = head;
+            head = head.next;
+            head.prev.next = head.prev = null;
+            return removedNode.item;
+        }
+        Node<T> currentNode = getNodeByIndex(index);
+        currentNode.prev.next = currentNode.next;
+        currentNode.next.prev = currentNode.prev;
+        size--;
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        if (!isValid(getIndexByValue(object))) {
+            return false;
+        }
+        remove(getIndexByValue(object));
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        return index > size / 2 ? startWithTail(index) : startWithHead(index);
+    }
+
+    private Node<T> startWithHead(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private Node<T> startWithTail(int index) {
+        Node<T> currentNode = tail;
+        for (int i = size - 1; i > index; i--) {
+            currentNode = currentNode.prev;
+        }
+        return currentNode;
+    }
```

Comment:
I think you can do in in one method. Just put into method index and  decide how to move inside method

Reply:
let's leave it as it is